ospf6d: Fix summary LSA removal

gromit1811 opened the pull request:
Finally, a first attempt to fix #16197

The main fixes are from the following 2 commits (other commits are mostly cleanup/optimization):

commit 9eb04927ae99392f0030ab22efb820156d4ae8d1

    ospf6d: Fix old summary route lookup
    
    When orignating summaries, we perform a lookup to find possible previous
    summaries. Don't just compare the prefix in that case, but also route and
    path properties: We might have several summary entries for the same
    destination (e.g. with a intra- and inter-area path) at the same time (at
    least temporarily) and we need to deal with the right one.


commit 6f8df0c25419c4f632188919b7087cf6f19c479e

    ospf6d: Fix summary LSA removal by using ospf6_abr_task()
    
    Previously, we checked each individual route to see whether a summary LSA
    needs to be originated or purged. The problem is that this decision can't be
    made reliably by looking at a single route, because there might be several
    routes for a destination (e.g. one with a intra- and one with an inter-area
    path) and one might cause origination of a summary LSA while the other
    causes it to be purged. This causes unstable and/or missing LSAs as
    described in https://github.com/FRRouting/frr/issues/16197
    
    Instead, ospf6_abr_originate_summary_to_area() now only originates new LSAs
    or approves existing ones. Removal is handled in ospf6_abr_task() after the
    whole routing table has been iterated. LSAs which have not been approved by
    any call to ospf6_abr_originate_summary_to_area() are then removed. This
    matches the behaviour of OSPFv2.


Note that this isn't finished/complete yet and is intended mainly for feedback or to stop me in case I'm moving in a completely wrong direction :wink:

Open issues:

- [x] Check topotest failure: ospf6_gr_topo1
- [x] Check topotest failure: ospf6_topo2
- [x] Routing table hooks no longer make sense the way they're implemented (directly calling ospf6_abr_originate_summary_to_area()), just schedule ospf6_abr_task() instead?
- [x] Right now we actually check routes twice, once via hooks and once via ospf6_abr_task(). Doesn't seem to hurt, but isn't really efficient.
- [x] Check whether there are other cases where we might have to schedule ospf6_abr_task() to immediately remove a summary LSA.
- [x] ospf6_abr_task() is hidden in ospf6_nssa.c even though it is not at all related to NSSA. Guess it was put there when adding NSSA support and that needed a small part of the ospf6_abr_task() functionality. Move non-NSSA functions from ospf6_nssa.c to ospf6_abr.c.
- [x] Rebase to master

donaldsharp commented:
Hey can we submit the PR against master not 10.2?  All code must go into master first.

gromit1811 commented:
Latest update fixes the memory corruption triggered by ospf6_gr_topo1. Still a few TODOs left.

gromit1811 commented:
Yay, all ospf6 topotests work now

Jafaral commented:
Thank you @gromit1811

gromit1811 commented:
@donaldsharp Rebased from 10.2 to master now but that seems to have confused CI completely. No, there are not that many commits in this PR and it also shouldn't conflict with master. What should I do? Close this PR and create a new one? Or is it possible to reset the merge target (still says "... wants to merge 1,875 commits into FRRouting:stable/10.2" at the top)?

gromit1811 commented:
Found it :smile:

donaldsharp commented:
it looks ok to me now.

donaldsharp commented:
ci:rerun

gromit1811 commented:
CI and frrbot still seem to be slightly confused and this PR has an impressive amount of labels which it doesn't deserve, though :wink:

Anyway, IMO it's ready for review. Some review notes:
* Whenever ospf6d called called `ospf6_abr_originate_summary()` for a single route before, we now simply do a `ospf6_schedule_abr_task()` and process the whole routing table. So for large routing tables, this is significant overhead. However, it matches what ospfd does. There would be an intermediate alternative: Process only one prefix at a time, but all routes/paths for that prefix (instead of just one path as before). But that would require even more restructuring. Opinions?
* Hooks: Are the calls to `ospf6_schedule_abr_task()` from `ospf6_top_brouter_hook_add/remove()` useful at all? If I understand correctly, brouters are only modified during/after SPF and we run the ABR task after SPF anyway. `ospf6_top_route_hook_add/remove()` probably really needs them, though.
* I currently use a 5s ABR task scheduling delay like ospfd, but IMO this is a pretty long waiting time for a routing update. Should that be lowered? OTOH, most updates will be faster because the ABR task runs immediately after SPF completion.
* Should a scheduled ABR task timer be cancelled when executing the ABR task immediately (via direct call or ospf6_execute_abr_task())? ospfd doesn't do this either, but IMO this would be the right thing to do...

gromit1811 commented:
Rebased again to current master. Now CI looks much better and my local topotests were also successful (except for some BGP/SRv6/NHRP tests I really don't feel responsible for :wink:)
Shall I fix the frrbot and "Verify Source" style issues? Except for one change that's code I've not modified at all but only moved around. And I don't fully agree with the "WARNING: Unnecessary ftrace-like logging - prefer using ftrace" - ftrace might be nice, but it's much more effort to set up than a simple "debug ospf6 ...".

gromit1811 commented:
Fixed style issues found by frrbot. I did this in a separate commit so that the "move code around" commit doesn't move and modify code at the same time which would be more difficult to review.

I intentionally left the ospf6_abr_task() entry/exit log messages in which "[CI] Verify Source" complained about - ABR task may run asynchronously and it's important to understand when that happens when reading logs. And no, ftrace is not a suitable replacement as it's important to have those log messages together the other log messages and in sequence. 

I removed some other entry/exit messages from functions only called by ospf6_abr_task(), though - they're only called from there and ABR task already logs each individual step.

gromit1811 commented:
As expected, "Verify Source" still fails, but now with only one warning left triggered by the `ospf6_abr_task()` entry/exit logs I left in intentionally. Is there some kind of magic comment to override the warning or is it OK to just ignore it?

@riw777 Anything else I can do simplify review? I know this is a big PR (but not as big as the labels suggest :wink:), but I tried to split it up into reasonable commits which should be easier to review individually than the whole diff. And the largest commit only moves code, so that should be relatively easy as well...

riw777 reviewed (approved):
looks good

riw777 commented:
Not certain if we should merge past this or not:

```Report for ospf6_abr.c | 2 issues
===============================================
< WARNING: Unnecessary ftrace-like logging - prefer using ftrace
< #1499: FILE: /tmp/f1-2062112/ospf6_abr.c:1499:
Details at [https://ci1.netdef.org/browse/FRR-PULLREQ3-CHECKOUT-8861](https://ci1.netdef.org/browse/FRR-PULLREQ3-CHECKOUT-8861)

gromit1811 commented:
> Not certain if we should merge past this or not:
> 
> ```
> ===============================================
> < WARNING: Unnecessary ftrace-like logging - prefer using ftrace
> < #1499: FILE: /tmp/f1-2062112/ospf6_abr.c:1499:
> Details at [https://ci1.netdef.org/browse/FRR-PULLREQ3-CHECKOUT-8861](https://ci1.netdef.org/browse/FRR-PULLREQ3-CHECKOUT-8861)
> ```

Me neither. I fixed the other issues checkpatch was complaining about, but this debug message here was pretty important for me when analyzing logs, so I'd strongly prefer to leave it in. Of course, I can try to reformulate it to not trigger the warning. But IIRC, @donaldsharp also commented somewhere (Slack?) and questioned the usefulness of this warning as the ftrace recommendation doesn't really apply in this context.

gromit1811 commented:
Rephrased debug message to make checkpatch happy. Hopefully the final fix that allows to merge this :crossed_fingers:

gromit1811 commented:
ci:rerun

gromit1811 commented:
> ci has been running for two weeks? hmmm...
> 
> ci:rerun

@riw777 Is it just this PR with repeated CI failures totally unrelated to the actual changes in the PR, or is CI generally unstable ATM?

riw777 commented:
> The remaining 2 "Basic Tests" failures look unrelated, BGP/ISIS SRv6 problems. Shall I rebase to get a possible fix from current master or just ignore them?

I'm going to try to rerun just the failed tests to see what we can do
